File: src/converter/plugins/convertExpressionWithTypeArguments.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {convertNodeWithTypeArguments} from "./convertNodeWithTypeArguments.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertExpressionWithTypeArguments = createSimplePlugin((node, context, render) => {
    if (!ts.isExpressionWithTypeArguments(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    return render(node.expression) + convertNodeWithTypeArguments.render(node, context, render)
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertExpressionWithTypeArguments.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.isExpressionWithTypeArguments

val convertExpressionWithTypeArguments = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isExpressionWithTypeArguments(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    render(node.expression) + convertNodeWithTypeArguments.render(node, context, render)
}
